### PR TITLE
Adding TrueNAS ALUA support

### PR DESCRIPTION
**Pull request by voodoochief**

### CODE REVIEW

**Comment by stormi** [comment]
Hi! You need to sign off your commits on this repository, which means that the commit message must contain something like `Signed-off-by: Jaron Parsons <your.email@address>`

**Comment by stormi** [comment]
Looks good to me. Let's see what the maintainers will say about it!

**Comment by MarkSymsCtx** [comment]
What does the default multipath configuration from the base `device-mapper-multipath` package generate. We typically try to avoid needing to add things explicitly to the multipath.conf file unless the default handler does the wrong thing.

File: multipath/multipath.conf
Comment (olivierlambert) [inline multipath/multipath.conf]: Extra trailing `}` should be removed

Comment (voodoochief) [reply multipath/multipath.conf]: fixed

Comment (MarkSymsCtx) [reply multipath/multipath.conf:88]: Doesn't look like it is fixed.

**Review by MarkSymsCtx — CHANGES_REQUESTED** [review]

**Comment by olivierlambert** [comment]
@voodoochief do you need help on this PR?

**Comment by olivierlambert** [comment]
Okay so @voodoochief disappeared, so let's do it that ourselves. @Wescoeur can you finish the PR please?

**Comment by olivierlambert** [comment]
haha I somehow summoned you @voodoochief :wink: 

@MarkSymsCtx anything left preventing a merge on your side?

**Comment by voodoochief** [comment]
My apologies.  I have made the change in my branch. 
And some additional things we have found in testing to enhance the settings further.
I am git challenged.  so ill try to make it happy.

File: multipath/multipath.conf
Comment (olivierlambert) [inline multipath/multipath.conf:90]: indentation issue?

File: multipath/multipath.conf
Comment (olivierlambert) [inline multipath/multipath.conf:93]: same than my previous comment

File: multipath/multipath.conf
Comment (voodoochief) [reply multipath/multipath.conf:90]: bah..

**Comment by olivierlambert** [comment]
Okay so I think it's good to me :+1:

**Comment by MarkSymsCtx** [comment]
You need to signoff the commits. Probably need to do something like

- git rebase -i origin/master
- set all to edit
- git commit --amend -s
- git rebase --continue
(until done)
- git push -f ...
 
The change itself looks good now.

**Comment by jaronparsons** [comment]
Should be good I think

**Review by MarkSymsCtx — APPROVED** [review]

**Review by MarkSymsCtx — CHANGES_REQUESTED** [review]
You've modified the file to reformat more than just what you're adding. That's not acceptable. Please only change what you're adding.

**Comment by olivierlambert** [comment]
In other words, PRs need to be:

1. Atomic as possible
2. Matching their description

So let's focus on TrueNAS ALUA support for now, and then if other mods are wanted, create a PR per new supported config (eg QNAP etc.). Also, to make maintainer work easier, explaining the modification is important (especially when you are a storage vendor/specialist, and sometimes maintainers aren't, because nobody can know in details how each storage need to be setup). Eg for LIO modification, it needs to be clear why it would be changed (in a new PR).

Thanks again @jaronparsons !

**Comment by MarkSymsCtx** [comment]
> In other words, PRs need to be:
> 
> 1. Atomic as possible
> 2. Matching their description
> 
> So let's focus on TrueNAS ALUA support for now, and then if other mods are wanted, create a PR per new supported config (eg QNAP etc.). Also, to make maintainer work easier, explaining the modification is important (especially when you are a storage vendor/specialist, and sometimes maintainers aren't, because nobody can know in details how each storage need to be setup). Eg for LIO modification, it needs to be clear why it would be changed (in a new PR).
> 
> Thanks again @jaronparsons !

Yes, indeed, the QNAP and LIO stuff isn't new, that's already in master but this PR reformatted the configuration stanzas for LIO and QNAP instead of just adding the TrueNAS stuff.

**Comment by jaronparsons** [comment]
Ok Apologies again, I just pulled in the changes from master.  It was giving me the conflict message and I thought that meant I needed to match it up.  ill try again.

**Comment by stormi** [comment]
This is not far from looking good.

Here's how to make it perfect:
* Fix the indentation of your closing `}`: should be vertically aligned with the `d` of `device`. I know you aligned it exactly like the previous entry, but that one is not a very good example of indentation (see the rest of the file). Don't fix the previous entry, this would be for another PR.
* Add a newline after your closing `}` so that your changes are all additions. Your current change removes an empty line, which again is a formatting change that does not belong to this PR.
* Last but not least, let's squash all your commits into one single commit. This will be cleaner in the project's git history: 
  * git rebase -i origin/master
  * set the first line to `edit` and all other lines to `fixup`
  * git commit --amend
    * Change the commit message to "Add TrueNAS ALUA support" (see rule 5 in https://chris.beams.io/posts/git-commit/)
    * Make sure to keep the sign-off-by line
  * git rebase --continue
  * git log # to check the result
  * git push -f
  * now your PR only contains one commit, which only changes what you meant to change

**Comment by jaronparsons** [comment]
I removed the other changes, but I could not see the indention offset you referred to .. it looks correct here.. but i re typed it regardless hopefully that clears it up.  squashed the commits, although I had to do it another way, but it seems to have worked. and added the extra line before the last closed bracket.  does this look acceptable now?

**Comment by stormi** [comment]
Regarding the indentation, I know what's wrong: your additions are indented with spaces when the original file is indented with tabs (and a few spaces here and there). Github doesn't display that kind of detail in commit diffs apparently. Looking at the last commit that was merged that changed this file on master (adding lines all indented with 4 spaces instead of tabs), I wouldn't worry too much about that, despite my personal distaste for any mix of spaces and tabs :)
I may contribute a formatting fix after your PR has been merged.

There's another issue with the current state of your PR: it's based on an old version of the master branch. You can see this at https://github.com/voodoochief/sm: your master branch is "1 commit ahead, 65 commits behind xapi-project:master.". And since there have been changes to multipath.conf on xapi-project:master (the latest having been merged a few hours ago), your PR conflicts.

To fix it:
```
# Add upstream repo as a new remote in your git working copy
git remote add upstream git@github.com:xapi-project/sm.git
# Go to your local master branch if not on it already
git checkout master
# Rebase it on the latest upstream master branch
git rebase upstream/master
# Fix conflicts manually: retain upstream changes and add yours at the bottom.
# Then:
git add multipath/multipath.conf # mark the conflict as solved
# Finish the rebase
git rebase --continue # or is it git commit? I'll check and edit this post.
# Push your forked master branch again
git push -f
```

**Comment by jaronparsons** [comment]
Unfortunately when i follow these steps, I get the following..
The first command give no error or response even.

after issuing "git checkout master" i get--

Already on 'master'
Your branch is up to date with 'origin/master'.

then if I run the rebase command i get
fatal: invalid upstream 'upstream/master'

--
Jaron Parsons
IXsystems, Inc. <http://www.ixsystems.com>
Manager of IT
Office: (408) 943-4100 ext.351

**Comment by voodoochief** [comment]
Closing this.  Will create a new PR with correct account and just TrueNAS setting changes.

**Comment by Fohdeesha** [comment]
@voodoochief any updates here? It seems this PR is needed, as the current truenas entry in smapi is missing the no_path_retry setting, which is present in your PR here. We have encountered at least one customer in which this setting not being present caused storage issues during controller failover.

**Pull request closed without merging**
